### PR TITLE
add masking

### PR DESCRIPTION
**Pull request by lbaierreinio**
There are two cases we need to handle: Prepadding (PAD, PAD, ... , X1, ...) & postpadding (X1, ..., PAD, PAD).

In SQUAD, post-padding will correspond to the forward layer in the bidirectional MinGRU, and pre-padding will correspond to the backward layer.

After applying masking, a[i] = b[i] = 0 for padded tokens at index i regardless of whether we are doing pre or postpadding, and therefore h[i] = 0 as well. Crucially, in prepadding, the hidden states of the actual tokens will not be influenced by the hidden states of the padded tokens.

Then, in BiMinGRU, we ensure that we are also flipping the mask, so that it lines up with the padded tokens in the reversed case.

Since h[I] = 0 for all padded tokens, we ensure that they never contribute to the hidden states of the actual tokens

### CODE REVIEW

File: src/layers/BiMinGRU.py [hunk 1]
```diff
@@ -9,11 +9,12 @@ def __init__(self, dim_in, dim_hidden):
     self.backward_rnn = MinGRU(dim_in, dim_hidden)
     self.linear = nn.Linear(2 * dim_hidden, dim_hidden)
   
-  def forward(self, x):
```
Comment (lbaierreinio) [inline src/layers/BiMinGRU.py:12]: Not sure if .flip() is best practice or if there is a more efficient way to do it

File: src/layers/MinGRU.py [hunk 2]
```diff
@@ -61,25 +64,27 @@ def forward(self, x, h_prev=None, mask=None):
...
 
             if mask is not None:
-                mask = mask.unsqueeze(-1) # [batch_size, seq_len, 1]
-                # NOTE: because minGRU computes hidden states sequentially (unlike attention)
-                #       if suffices to mask out the irrelevant hidden states to 0
-                h = h.masked_fill(mask, 0)
+                mask = mask.unsqueeze(-1)
+                log_z = log_z.masked_fill(mask, 0)
+                log_tilde_h = log_tilde_h.masked_fill(mask, 0)
+                log_one_minus_z = log_one_minus_z.masked_fill(mask, 0)
```
Comment (dlrudwo1269) [inline src/layers/MinGRU.py:85]: If we perform masking here, I believe we should mask with `-inf` since we are operating in the log domain. We do a `.exp()` before returning in `parallel_log_scan` which means values that are set to 0 will become 1.

But on second thought, I am thinking maybe masking is not required here at all. In attention, masking is necessary, because every token can attend to every other token including padding. However, this is a sequential computation and if we flip the sequence for the reverse direction in `BiMinGRU`, this block only handles post-padded sequences. So padding tokens never affect the values of the tokens we care about.

File: src/layers/BiMinGRU.py [hunk 3]
```diff
@@ -9,11 +9,12 @@ def __init__(self, dim_in, dim_hidden):
...
+  def forward(self, x, mask):
     x_reversed = x.flip(dims=[1])
+    mask_reversed = mask.flip(dims=[1])
     out_forward = self.forward_rnn(x)
-    out_backward = self.backward_rnn(x_reversed)
-    concat = torch.cat((out_forward, out_backward), dim=2)
+    out_backward = self.backward_rnn(x_reversed, mask_reversed)
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/layers/BiMinGRU.py:16]: (a) yes, missed applying the mask in the fwd direction
(b) I think we can just have two named arguments mask & h_prev

File: src/layers/MinGRU.py (hunk 2 above)
Comment (dlrudwo1269) [reply src/layers/MinGRU.py:85]: Oh right yeah we need to mask. I was being confused.

In that case, I think this is good as it is 👍

**Review by dlrudwo1269 — APPROVED** [review]
LGTM
